dulwich/index.py

```diff
@@ -516,7 +516,12 @@ def blob_from_path_and_stat(fs_path, st):
         with open(fs_path, 'rb') as f:
             blob.data = f.read()
     else:
-        blob.data = os.readlink(fs_path)
+        if sys.platform == 'win32':
+            fs_encoding = sys.getfilesystemencoding()
+            blob.data = (os.readlink(fs_path.decode(fs_encoding))
+                         .encode(fs_encoding))
```

Comment:
On Mon, May 25, 2015 at 05:30:36AM -0700, Gary van der Merwe wrote:
> > @@ -516,7 +516,12 @@ def blob_from_path_and_stat(fs_path, st):
> >          with open(fs_path, 'rb') as f:
> >              blob.data = f.read()
> >      else:
> > -        blob.data = os.readlink(fs_path)
> > +        if sys.platform == 'win32':
> > +            fs_encoding = sys.getfilesystemencoding()
> > +            blob.data = (os.readlink(fs_path.decode(fs_encoding))
> > +                         .encode(fs_encoding))
> 
> Yes :-( . This is the same for other types of files, because methods such as `open` will implicitly decode the path when passed bytes, using `sys.getfilesystemencoding()`, which will be `'mbsc'`, as the encoding. This is the cause of #203. 
Can you require the caller pass in the tree_encoding and then pass in sys.getfilesystemencoding() ? At least that way the intent is clear.

Cheers,

Jelmer